src/ui/components/HostPermissions/index.js

```diff
@@ -0,0 +1,166 @@
+    // https://hg.mozilla.org/mozilla-central/raw-file/tip/browser/locales/en-US/chrome/browser/browser.properties
+    switch (messageType) {
+      case 'allUrls':
+        return this.i18n.gettext('Access your data for all websites');
+      case 'domain':
+        if (tooMany) {
+          return this.i18n.sprintf(this.i18n.ngettext(
+            'Access your data in %(param)s other domain',
+            'Access your data in %(param)s other domains',
+            param), { param: this.i18n.formatNumber(param) }
+          );
+        }
+        return this.i18n.sprintf(
+          this.i18n.gettext('Access your data for sites in the %(param)s domain'),
+          { param }
+        );
+      case 'site':
+        if (tooMany) {
+          return this.i18n.sprintf(this.i18n.ngettext(
+            'Access your data on %(param)s other site',
+            'Access your data on %(param)s other sites',
+            param), { param: this.i18n.formatNumber(param) }
+          );
+        }
+        return this.i18n.sprintf(
+          this.i18n.gettext('Access your data for %(param)s'),
+          { param }
+        );
+      default:
+        return '';
+    }
+  }
+
+  i18n: I18nType;
+
+  // Generates Permission components for a list of host permissions. If we have 4 or
+  // fewer, display them all, otherwise display the first 3 followed by an item
+  // that says "...plus N others".
+  generateHostPermissions({
+    permissions, messageType,
+  }: GenerateHostPermissionsParams): Array<React.Element<typeof Permission>> {
+    const hostPermissions = [];
+    for (const item of permissions.slice(0, 4)) {
+      // Add individual Permission components for the first 4 host permissions.
+      hostPermissions.push(
+        <Permission
+          type="hostPermission"
+          description={this.getPermissionString({ messageType, param: item })}
+          key={item}
+        />
+      );
+    }
+    if (permissions.length > 4) {
+      // Replace the final individual permission with a "too many" permission.
+      hostPermissions[3] = (<Permission
+        type="hostPermission"
+        description={this.getPermissionString(
+          { messageType, param: permissions.length - 3, tooMany: true }
+        )}
+        key={messageType}
+      />);
+    }
+    return hostPermissions;
+  }
+
+  render() {
+    const { permissions } = this.props;
+    const hostPermissions = [];
+
+    // Group permissions into "site" and "domain" permissions. If any
+    // "all urls" permissions are found, break as we'll only need one
+    // host permissions message.
+    let allUrls = false;
+    const wildcards = [];
+    const sites = [];
+    for (const permission of permissions) {
+      if (permission === '<all_urls>') {
+        allUrls = true;
+        break;
+      }
+      if (permission.startsWith('moz-extension:')) {
+        continue;
+      }
+      const match = /^[a-z*]+:\/\/([^/]+)\//.exec(permission);
+      if (!match) {
```

Comment:
When would this happen? These permissions are all validated by AMO, right? This seems like it should be more than an error that we continue with.

Reply:
This can happen, but it's highly unlikely. One example of a pattern that makes it this far is the pattern `*`. 

If the pattern is invalid and wasn't caught by AMO it also won't be applied by the extension, so I think ignoring is fine in this case. This is similar code to what Firefox does when displaying permissions to a user for the dialog.